guillotina_gcloudstorage/storage.py

```diff
@@ -204,34 +204,47 @@ def should_clean(self, file):
         else:
             raise AttributeError('No valid uri')
 
-    async def _append(self, dm, data, offset):
-        if dm.size:
+    async def _append(self, dm, data, offset, is_last):
+        if dm.size is not None:
             size = dm.size
+        elif is_last:
+            size = len(data) + offset
         else:
             # assuming size will come eventually
             size = '*'
-        content_range = 'bytes {init}-{chunk}/{total}'.format(
-            init=offset,
-            chunk=offset + len(data) - 1,
-            total=size)
+        headers = {
+            'Content-Length': str(len(data)),
+            'Content-Type': to_str(dm.content_type),
+        }
+        if len(data) != size:
+            content_range = 'bytes {init}-{chunk}/{total}'.format(
+                init=offset,
+                chunk=offset + len(data) - 1,
+                total=size)
+            headers['Content-Range'] = content_range
+
         util = get_utility(IGCloudBlobStore)
         async with util.session.put(
                 dm.get('resumable_uri'),
-                headers={
-                    'Content-Length': str(len(data)),
-                    'Content-Type': to_str(dm.content_type),
-                    'Content-Range': content_range
-                },
+                headers=headers,
                 data=data) as call:
             text = await call.text()  # noqa
             if call.status not in [200, 201, 308]:
                 log.error(text)
             return call
 
+    async def _async_gen_lookahead(self, gen):
+        async for prev in gen:
```

Comment:
maybe instead of nested we can use something with `next(...)`?

This looks a bit wonky.

Also, I think it can to moved into the `append` method if we do it this way.

You'd also catch the `StopIteration` exception then.